src/lib/party.ts

```diff
@@ -1052,42 +1108,44 @@ const CallForest = {
 
   // similar to Mina_base.Parties.Call_forest.accumulate_hashes
   // hashes a party's children (and their children, and ...) to compute the `calls` field of ZkappPublicInput
-  hashChildren(parent: Party) {
+  hashChildren({ children }: Party): Field {
+    // only compute calls if it's not there yet --
+    // this gives us the flexibility to witness a specific layout of parties
+    if (children.calls !== undefined) return children.calls;
     let stackHash = CallForest.emptyHash();
-    for (let { party, calls } of parent.children.reverse()) {
-      // only compute calls if it's not there yet --
-      // this gives us the flexibility to witness only direct children of a zkApp
-      calls ??= CallForest.hashChildren(party);
+    for (let party of [...children.parties].reverse()) {
+      let calls = CallForest.hashChildren(party);
       let nodeHash = hashWithPrefix(prefixes.partyNode, [party.hash(), calls]);
-      stackHash = hashWithPrefix(prefixes.partyCons, [nodeHash, stackHash]);
+      // stackHash = hashWithPrefix(prefixes.partyCons, [nodeHash, stackHash]);
+      let newHash = hashWithPrefix(prefixes.partyCons, [nodeHash, stackHash]);
+      // skip party if it's a dummy
+      stackHash = Circuit.if(party.isDummy(), stackHash, newHash);
     }
     return stackHash;
   },
+
+  forEach(updates: Party[], callback: (update: Party) => void) {
+    for (let update of updates) {
+      callback(update);
+      CallForest.forEach(update.children.parties, callback);
+    }
+  },
 };
 
 function createChildParty(
   parent: Party,
   childAddress: PublicKey,
-  options?: {
-    caller?: Field;
-    tokenId?: Field;
-    useFullCommitment?: Bool;
-  }
+  tokenId?: Field
```

Comment:
simplified the options here to make it have the same `publicKey` + `tokenId` interface as almost all other methods that create a party